x/config/config.go

```diff
@@ -0,0 +1,117 @@
+package config
+
+import (
+	"fmt"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/transport"
+)
+
+// YARPC TODO
+type YARPC struct {
+	Name       string
+	Inbounds   []InboundConfig
+	Outbounds  []OutboundConfig
+	Transports []TransportConfig
+}
+
+// Build TODO
+func (c *YARPC) Build() (yarpc.Config, error) {
+	cfg := yarpc.Config{Name: c.Name, Outbounds: make(yarpc.Outbounds)}
+
+	transports := make(map[string]transport.Transport)
+	for _, tcfg := range c.Transports {
+		t, err := tcfg.Builder.BuildTransport()
+		if err != nil {
+			return cfg, fmt.Errorf("failed to build transport %q: %v", tcfg.Name, err)
+		}
+		transports[tcfg.Name] = t
+	}
+
+	for _, icfg := range c.Inbounds {
+		tname := icfg.TransportName
+		// TODO: error if transport not found in map
+		inbound, err := icfg.Builder.BuildInbound(transports[tname])
+		if err != nil {
+			return cfg, fmt.Errorf("failed to build inbound %q: %v", tname, err)
+		}
+		cfg.Inbounds = append(cfg.Inbounds, inbound)
+	}
+
+	for _, ocfg := range c.Outbounds {
+		outbounds := transport.Outbounds{ServiceName: ocfg.Service}
+		if ocfg.Oneway != nil {
+			tname := ocfg.Oneway.TransportName
+			// TODO: error if transport not found in map
+			oneway, err := ocfg.Oneway.Builder.BuildOnewayOutbound(transports[tname])
+			if err != nil {
+				return cfg, fmt.Errorf("failed to build oneway outbound %q: %v", ocfg.Name, err)
+			}
+			outbounds.Oneway = oneway
+		}
+		if ocfg.Unary != nil {
+			tname := ocfg.Unary.TransportName
+			// TODO: error if transport not found in map
+			unary, err := ocfg.Unary.Builder.BuildUnaryOutbound(transports[tname])
+			if err != nil {
+				return cfg, fmt.Errorf("failed to build unary outbound %q: %v", ocfg.Name, err)
+			}
+			outbounds.Unary = unary
+		}
+		cfg.Outbounds[ocfg.Name] = outbounds
+	}
+
+	return cfg, nil
+}
+
+// TransportConfig TODO
+type TransportConfig struct {
+	Name    string
+	Builder TransportBuilder
+}
+
+// InboundConfig TODO
+type InboundConfig struct {
+	TransportName string
+	Builder       InboundBuilder
+}
+
+// OutboundConfig TODO
+type OutboundConfig struct {
+	Name    string
+	Service string
+	Unary   *UnaryOutboundConfig
```

Comment:
We can/should provide `String()` functions for the types defined here but unless the injected configuration types also implement Stringer, we're going to end up with unreadable output.

Even with Stringer on all types, everything will be a single-line blob. If we're not okay with that, we can rely on something like [`kr/pretty`](https://github.com/kr/pretty) to output readable configuration dumps.